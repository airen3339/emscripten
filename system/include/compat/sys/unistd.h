#include <unistd.h>
#include <random.h> // for getentropy()
