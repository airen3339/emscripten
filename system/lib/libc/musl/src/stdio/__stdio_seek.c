#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
