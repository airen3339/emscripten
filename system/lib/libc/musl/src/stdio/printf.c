#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

