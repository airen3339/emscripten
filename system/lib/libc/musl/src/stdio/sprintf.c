#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

