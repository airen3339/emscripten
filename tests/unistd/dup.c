#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>

int main() {
  int f, f2, f3;
  off_t offset;

  EM_ASM(
    FS.mkdir('/working');
#if NODEFS
    FS.mount(NODEFS, { root: '.' }, '/working');
#endif
    FS.writeFile('/working/file', 'test');
  );

  printf("DUP\n");
  f = open("/working/file", O_RDONLY);
  f2 = open("/working/file", O_RDONLY);
  f3 = dup(f);
  printf("errno: %d\n", errno);
  printf("f: %d\n", f != f2 && f != f3);
  printf("f2,f3: %d\n", f2 != f3);

  // dup()ed file descriptors should share all flags and even seek position
  offset = lseek(f3, 0, SEEK_CUR);
  printf("1. f3 offset was %d.    Should be 0\n", (int)offset);
  offset = lseek(f, 1, SEEK_SET);
  printf("2. f  offset set to %d. Should be 1\n", (int)offset);
  offset = lseek(f2, 2, SEEK_SET);
  printf("3. f2 offset set to %d. Should be 2\n", (int)offset);
  offset = lseek(f, 0, SEEK_CUR);
  printf("4. f  offset now is %d. Should be 1\n", (int)offset);
  offset = lseek(f2, 0, SEEK_CUR);
  printf("5. f2 offset now is %d. Should be 2\n", (int)offset);
  offset = lseek(f3, 0, SEEK_CUR);
  printf("6. f3 offset now is %d. Should be 1 (and not 0)\n", (int)offset);
  offset = lseek(f3, 3, SEEK_SET);
  printf("7. f3 offset set to %d. Should be 3\n", (int)offset);
  offset = lseek(f, 0, SEEK_CUR);
  printf("8. f  offset now is %d. Should be 3 (and not 1)\n", (int)offset);

  printf("close(f1): %d\n", close(f));
  printf("close(f2): %d\n", close(f2));
  printf("close(f3): %d\n", close(f3));
  printf("\n");
  errno = 0;

  printf("DUP2\n");
  f = open("/", O_RDONLY);
  f2 = open("/", O_RDONLY);
  f3 = dup2(f, f2);
  printf("errno: %d\n", errno);
  printf("f: %d\n", f != f2 && f != f3);
  printf("f2,f3: %d\n", f2 == f3);
  printf("close(f1): %d\n", close(f));
  printf("close(f2): %d\n", close(f2));
  printf("close(f3): %d\n", close(f3));
  errno = 0;

  return 0;
}
