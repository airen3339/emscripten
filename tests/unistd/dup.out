DUP
errno: 0
f: 4
f2: 5
f3: 6
close(f1): 0
close(f2): 0
close(f3): 0

DUP2
errno: 0
f: 7
f2: 8
f3: 8
close(f1): 0
close(f2): 0
close(f3): -1
